libraries/rush-lib/src/logic/operations/BuildPlanPlugin.ts

```diff
@@ -0,0 +1,298 @@
+
+const PLUGIN_NAME: 'BuildPlanPlugin' = 'BuildPlanPlugin';
+
+interface IBuildPlanOperationCacheContext {
+  cacheDisabledReason: IOperationBuildCacheContext['cacheDisabledReason'];
+}
+
+interface ICobuildPlan {
+  summary: {
+    maxWidth: number;
+    maxDepth: number;
+    numberOfNodesPerDepth: number[];
+  };
+  executionPlan: Operation[];
+  clusters: Set<Operation>[];
+  buildCacheByOperation: Map<Operation, IBuildPlanOperationCacheContext>;
+  clusterByOperation: Map<Operation, Set<Operation>>;
+}
+
+export class BuildPlanPlugin implements IPhasedCommandPlugin {
+  private readonly _terminal: ITerminal;
+
+  public constructor(terminal: ITerminal) {
+    this._terminal = terminal;
+  }
+
+  public apply(hooks: PhasedCommandHooks): void {
+    const terminal: ITerminal = this._terminal;
+    hooks.beforeExecuteOperations.tapPromise(PLUGIN_NAME, createBuildPlan);
+
+    async function createBuildPlan(
+      recordByOperation: Map<Operation, IOperationExecutionResult>,
+      context: IExecuteOperationsContext
+    ): Promise<void> {
+      const { projectConfigurations, projectChangeAnalyzer } = context;
+      const disjointSet: DisjointSet<Operation> = new DisjointSet<Operation>();
+      const operations: IterableIterator<Operation> = recordByOperation.keys();
+      for (const operation of operations) {
+        disjointSet.add(operation);
+      }
+      const buildCacheByOperation: Map<Operation, IBuildPlanOperationCacheContext> = new Map<
+        Operation,
+        IBuildPlanOperationCacheContext
+      >();
+      for (const operation of operations) {
+        const { associatedProject, associatedPhase } = operation;
+        if (associatedProject && associatedPhase) {
+          const projectConfiguration: RushProjectConfiguration | undefined =
+            projectConfigurations.get(associatedProject);
+          const fileHashes: Map<string, string> | undefined =
+            await projectChangeAnalyzer._tryGetProjectDependenciesAsync(associatedProject, terminal);
+          const cacheDisabledReason: string | undefined = projectConfiguration
+            ? projectConfiguration.getCacheDisabledReason(fileHashes!.keys(), associatedPhase.name, operation)
+            : `Project does not have a ${RushConstants.rushProjectConfigFilename} configuration file, ` +
+              'or one provided by a rig, so it does not support caching.';
+          buildCacheByOperation.set(operation, { cacheDisabledReason });
+        }
+      }
+      clusterOperations(disjointSet, buildCacheByOperation);
+      const buildPlan: ICobuildPlan = createCobuildPlan(disjointSet, terminal, buildCacheByOperation);
+      logCobuildBuildPlan(buildPlan, terminal);
+    }
+  }
+}
+
+function generateCobuildPlanSummary(operations: Operation[], terminal: ITerminal): ICobuildPlan['summary'] {
+  const noOpOperations: Set<Operation> = new Set(operations.filter((e) => e.runner?.isNoOp));
```

Comment:
It's probably cheaper to just define:
```ts
function isNoOp(operation: Operation): boolean {
  return operation?.runner?.isNoOp === true;
}
```
And call that instead of creating the set. I expect the lookup cost of the property access is negligible and I think it's easier to read calling the function.

Reply:
Heck, feel free to add a `get isNoOp() { return !!this.runner?.isNoOp; }` on `Operation`.

Reply:
Updated!